src/machines/activity_state_machine.rs

```diff
@@ -417,9 +537,73 @@ pub(super) struct TimedOut {}
 #[derive(Default, Clone)]
 pub(super) struct Canceled {}
 
+fn create_request_cancel_activity_task_command(
+    dat: SharedState,
+    next_state: ActivityMachineState,
+) -> TransitionResult<ActivityMachine> {
+    // createRequestCancelActivityTaskCommand
+    let cmd = Command {
+        command_type: CommandType::RequestCancelActivityTask as i32,
+        attributes: Some(
+            RequestCancelActivity {
+                scheduled_event_id: dat.scheduled_event_id,
+                activity_id: dat.attrs.activity_id,
+            }
+            .into(),
+        ),
+    };
+    ActivityMachineTransition::ok(
+        vec![ActivityMachineCommand::RequestCancellation(cmd)],
+        next_state,
+    )
+}
+
+/// Notifies lang side that activity has been cancelled by sending a failure with cancelled failure as a cause.
+/// Optional cancelled_event, if passed, is used to supply event IDs.
+fn notify_canceled(
+    dat: SharedState,
+    canceled_event: Option<ActivityTaskCanceledEventAttributes>,
+    next_state: ActivityMachineState,
+) -> TransitionResult<ActivityMachine> {
+    let (scheduled_event_id, started_event_id) = match canceled_event {
+        None => (dat.scheduled_event_id, dat.started_event_id),
+        Some(e) => (e.scheduled_event_id, e.scheduled_event_id),
+    };
+    let cancelled_failure = Failure {
+        source: "CoreSDK".to_string(),
+        failure_info: Some(FailureInfo::CanceledFailureInfo(CanceledFailureInfo {
+            details: None,
+        })),
+        ..Default::default()
+    };
+    let activity_failure_info = ActivityFailureInfo {
+        activity_id: dat.attrs.activity_id.to_string(),
+        activity_type: Some(ActivityType {
+            name: dat.attrs.activity_type.to_string(),
+        }),
+        scheduled_event_id,
+        started_event_id,
+        identity: "workflow".to_string(),
+        retry_state: RetryState::Unspecified as i32,
+    };
+    ActivityMachineTransition::ok_shared(
+        vec![ActivityMachineCommand::Fail(Some(Failure {
```

Comment:
I think we should extract the construction of this big `Failure` struct to it's own function so that it de-clutters the important logic here.

Reply:
Extracted and called new_cancellation_failure